### DISCUSSION TITLE
Warhammer: Shadow of the Horned Rat

### DISCUSSION THREAD

**Discussion by aqrit** [danoon2/Boxedwine#34] [Game/App Help] [answered]
Warhammer: Shadow of the Horned Rat

A boxedwine build, of the current master branch (f8a3521), for windows x86 is unable to reach the first battle.
(However, the game does work with boxedwine builds for: windows x64, linux x86, and linux x64.)

The game doesn't crash -- and is still responsive -- but it never draws the traveling animation (red dots on the map screen) and never starts the actual battle.
 
Tested the Win95 versions of v1.1 and the gog.com release (v1.0?). 
The Win3.1 verision was not tested. 
The v1.1 installer was also not tested.

====

~~The smacker movies in game version 1.1 "appear" to have color palette issues in boxedwine,
but hacked-up[*] videos work using regular `linux\Wine`.~~

The movies in the GOG release work fine in boxedwine.
GOG replaced `SMACKW32.DLL` with `OMNI000R.DLL`, I'm not sure what that is about.

[*] The game checks for 256 color depth... but it doesn't actually need it.
Nopping out that check and running it in 32bpp appears to works fine under `kubuntu 20.04 / Wine 5.0.3ubuntu` (or w/e) without screwing around with xserver.
On native windows, I'd also hot-patch `ChangeDisplaySettingExA` to do nothing because of all the compatibility shims.

```
File: WHSHR.EXE (v1.1)
Description: Skip check for 8bpp display.
File Offset: 0x00041833
Patch Bytes: 0x90 0x90 0x90 0x90 0x90 0x90
Original Bytes: 0x0F 0x85 0x0C 0x00 0x00 0x00

File: WHSHR.EXE (v1.1)
Description: Force 32bpp display.
File Offset: 0x00042EC0
Patch Bytes: 0x6A 0x20
Original Bytes: 0x6A 0x08
```
====

Installing the GOG.com Release in BoxedWine

1. Download the offline copy of the installer from gog.com
2. Download https://constexpr.org/innoextract/
3. Drag & Drop the installer onto innoextract
4. Copy the extracted `WARFB` folder to your `.wine\drive_c`
5. Copy the `RL` folder from `__support\sys32` to `.wine\drive_c\windows\system32`
6. Copy the `RLDDF.DLL` file from `__support\sys32` to `.wine\drive_c\windows\system32`
7. Import the following registry entries using boxedwine's regedit.exe

```
Windows Registry Editor Version 5.00

[HKEY_CURRENT_USER\Software\Mindscape]

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB]
@=""
"Options"="3,1,1,1,1,1,1,1"

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB\InstallDir]
@="C:\\WARFB"

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB\InstallOption]
@="Maximum"

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB\Language]
@="ENG"

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB\LocalDir]
@="C:\\WARFB\\FILE"

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB\RemoteDir]
@="C:\\WARFB\\REMOTE"

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB\SaveDir]
@="C:\\WARFB\\TEMP"

[HKEY_CURRENT_USER\Software\Mindscape\WarHammerFB\UpdateDir]
@="C:\\WARFB\\Update"
```

danoon, if you care to take a peek, I PM'ed a GOG gift code for this game on vogons.

**Comment by danoon2** [comment]
Boxedwine also supports the idea of returning that it is running in 256 color mode, even though it doesn't.  Wine GDI will automatically convert it.  On the command line this is -bpp 8 or in the UI for Boxedwine, when you look at the game options in the Containers tab, you can select 8-bit for "Bits per Pixel:"

As for the the not starting the battle (no red dots showing the travel), I will look into it.

For the x64 build, did the GOG installer not work for you?  They often have errors at the end of the installer, but for the most part GOG installers seem to work even with the error msg boxes.

**Reply by aqrit** [reply]
The v1.1 movies:
don't work under boxedwine in 8bpp.
don't work under boxedwine in 32bpp w/patch.
do work under regular wine in 32bpp w/patch.
Unknown whether they work under regular wine in 8bpp.


Under regular wine, the GOG installer doesn't get the Reality Labs middleware into the system32/syswow64 folder.
Under boxedwine, the installer never shows a gui (I don't know if the installer requires mono or some such).

**Comment by danoon2** [comment, marked as answer]
Thanks for posting that you tried the Linux x86 build and that it worked for you.  That quickly led me to an area to research.  It turns out that I never enabled the JIT for the build.sh script for Linux even though it should work.  For the last 2 years I've really just been focused on x64 and ARMv8.

I think I did find the issue for this and it was in the JIT.  It was an interesting case that my unit tests don't currently cover where a single block of code contains two instructions that set flags and the 2nd one doesn't clobber the first.  I think this can only happen to a couple of instructions like inc/dec.  In this case there was a shift which set up some lazy flags and then a decrement instruction came after it.  There was a bug in this case where the register containing the decrement result was clobbered before committed to memory.

Go ahead and try the master branch to see if it fixes your issue with not getting to the battle.

**Reply by aqrit** [reply]
Nice job, it works!

I'll poke around with the movies and see if I can isolate the issue,

**Reply by aqrit** [reply]
Patching the game to force 32 bpp color fixes the smacker movies under BoxedWine. So all I can conclude is that BoxedWine works too well :-)

**Reply by danoon2** [reply]
Glad to hear you got it working.  Older games, especially around that time of 1995/96, can be pretty challenging to run.
